language/go/generate.go

```diff
@@ -427,53 +478,125 @@ func (g *generator) generateProto(mode proto.Mode, target protoTarget, importPat
-		importPath := InferImportPath(g.c, g.rel)
+		if importPath == "" {
+			importPath = InferImportPath(g.c, g.rel)
+		}
 		protoName = proto.RuleName(importPath)
 	}
 	goProtoName := strings.TrimSuffix(protoName, "_proto") + goProtoSuffix
 	visibility := g.commonVisibility(importPath)
 
 	if mode == proto.LegacyMode {
 		filegroup := rule.NewRule("filegroup", filegroupName)
-		if target.sources.isEmpty() {
+		if targets[0].sources.isEmpty() {
 			return "", []*rule.Rule{filegroup}
 		}
-		filegroup.SetAttr("srcs", target.sources.build())
+		filegroup.SetAttr("srcs", targets[0].sources.build())
 		if g.shouldSetVisibility {
 			filegroup.SetAttr("visibility", visibility)
 		}
 		return "", []*rule.Rule{filegroup}
 	}
 
-	if target.sources.isEmpty() {
+	var atLeastOneTargetHasSources bool
+	for _, target := range targets {
+		if !target.sources.isEmpty() {
+			atLeastOneTargetHasSources = true
+			break
+		}
+	}
+
+	if !atLeastOneTargetHasSources {
 		return "", []*rule.Rule{
 			rule.NewRule("filegroup", filegroupName),
 			rule.NewRule("go_proto_library", goProtoName),
 		}
 	}
 
 	goProtoLibrary := rule.NewRule("go_proto_library", goProtoName)
-	goProtoLibrary.SetAttr("proto", ":"+protoName)
+	if len(targets) == 1 {
+		goProtoLibrary.SetAttr("proto", ":"+protoName)
+	} else {
+		// generate protoNames with prefix ":"
+		protoNames := make([]string, len(targets))
+		for i := range targets {
+			protoNames[i] = ":" + targets[i].name
+		}
+		goProtoLibrary.SetAttr("protos", protoNames)
+	}
+
 	g.setImportAttrs(goProtoLibrary, importPath)
-	if target.hasServices {
+	var atLeastOneTargetHasServices bool
+	for _, target := range targets {
+		if target.hasServices {
+			atLeastOneTargetHasServices = true
+			break
+		}
+	}
+	if atLeastOneTargetHasServices {
 		goProtoLibrary.SetAttr("compilers", gc.goGrpcCompilers)
 	} else if gc.goProtoCompilersSet {
 		goProtoLibrary.SetAttr("compilers", gc.goProtoCompilers)
 	}
 	if g.shouldSetVisibility {
 		goProtoLibrary.SetAttr("visibility", visibility)
 	}
-	goProtoLibrary.SetPrivateAttr(config.GazelleImportsKey, target.imports.build())
+	if len(targets) == 1 {
+		goProtoLibrary.SetPrivateAttr(config.GazelleImportsKey, targets[0].imports.build())
+	} else {
+		protoSources := make(map[string]bool)
+		for _, target := range targets {
+			for _, src := range target.sources.build().Generic {
+				// if src starts with the repo root plus a slash, trim it
+				if strings.HasPrefix(src, g.c.RepoRoot+"/") {
+					src = src[len(g.c.RepoRoot)+1:]
+				}
+				protoSources[src] = true
+			}
+		}
+		var combinedImports platformStringsBuilder
+		for _, target := range targets {
+			builtImports := target.imports.build()
+			// handle Generics
+			for _, genericString := range builtImports.Generic {
+				// if the generic string is not in the set of generic src strings, add it
+				if _, ok := protoSources[genericString]; !ok {
+					combinedImports.addGenericString(genericString)
+				}
+			}
+			//// handle OS
```

Comment:
Do we need OS/Arch/Platform strings? I guess proto imports are all independent of OS/Arch/Platform?

Reply:
Unsure. I was pattern matching and just trying to combine everything into a single platformStringsBuilder